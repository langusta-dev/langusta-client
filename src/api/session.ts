import { rest } from '~/helpers/api'

import type { AuthPayload } from '~/types/session'

export const logIn = async (data: AuthPayload) => {
  const response = await rest.post<{ token: string }>('/auth/login', data)
  return { ...response, data: response.data?.token || null }
}

export const register = async (data: AuthPayload) => {
  const response = await rest.post<{ token: string }>('/auth/register', data)
  return { ...response, data: response.data?.token || null }
}
