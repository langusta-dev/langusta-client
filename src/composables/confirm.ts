import type { ConfirmPayload } from '~/types/confirm'

const confirmEventBus = useEventBus<Partial<ConfirmPayload>>('confirm')

export const showConfirm = (payload: Partial<ConfirmPayload>) => {
  confirmEventBus.emit(payload)
}

export const addConfirmListener = (
  confirmListener: (payload: Partial<ConfirmPayload>) => void
) => confirmEventBus.on(confirmListener)
