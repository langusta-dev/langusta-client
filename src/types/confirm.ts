export interface ConfirmPayload {
  msg: string;
  confirmMsg: string;
  cancelMsg: string;
  cb: (() => void) | null;
}
