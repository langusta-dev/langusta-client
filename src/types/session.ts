export interface LogInPayload {
  username: string;
  password: string;
}

export interface RegisterPayload {
  email: string;
  username: string;
  firstname: string;
  lastname: string;
  password: string;
}
