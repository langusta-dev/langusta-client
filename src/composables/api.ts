import axios, { AxiosError } from 'axios';

import { HttpStatusCode } from '~/types/api';

import { jwt, unsetJwt } from '~/composables/jwt';
import { startNProgress, stopNProgress } from '~/composables/nprogress';

import { isOnline } from './online';

import type { AxiosRequestConfig, AxiosInstance, AxiosResponse } from 'axios';

type RequestResponse<T> = AxiosResponse<T | null>;

type RequestFn = <T = unknown>(url: string) => Promise<RequestResponse<T>>;

type RequestFnWithData = <T = unknown, D = unknown>(
  url: string,
  data?: D
) => Promise<RequestResponse<T>>;

type RestInstance = Record<'get' | 'delete', RequestFn> &
  Record<'post' | 'put' | 'patch', RequestFnWithData> &
  AxiosInstance;

class ClientOfflineError extends Error {
  constructor() {
    super('Client is offline');
  }
}

const CONFIG: AxiosRequestConfig = {
  baseURL: import.meta.env.VITE_API_BASE_URL,
  timeout: 10_000,
  withCredentials: true,
};

const rest = axios.create(CONFIG) as RestInstance & { silent: AxiosInstance };
rest.silent = axios.create(CONFIG) as RestInstance;
rest.silent.defaults.headers = rest.defaults.headers;

const handleClientOffline = (
  request: AxiosRequestConfig
): AxiosRequestConfig => {
  if (!isOnline.value) {
    throw new ClientOfflineError();
  }

  return request;
};

const parseResponse = (response: AxiosResponse): AxiosResponse => {
  if (response.status === HttpStatusCode.Unauthorized) {
    unsetJwt();
  }

  return { ...response, data: response.data ?? null };
};

const handleErrors = (error: Error | AxiosError): AxiosResponse => {
  if (error instanceof ClientOfflineError) {
    return {
      status: HttpStatusCode.ClientOffline,
      statusText: 'Client is offline',
      data: null,
      headers: {},
      config: CONFIG,
    };
  }

  if (error instanceof AxiosError) {
    return {
      status: error.response?.status || HttpStatusCode.UnknownError,
      statusText: error.response?.statusText || error.message,
      data: null,
      headers: error.response?.headers || {},
      config: CONFIG,
    };
  }

  return {
    status: HttpStatusCode.UnknownError,
    statusText: error.message,
    data: null,
    headers: {},
    config: CONFIG,
  };
};

rest.interceptors.request.use(handleClientOffline);
rest.silent.interceptors.request.use(handleClientOffline);

rest.interceptors.response.use(parseResponse, handleErrors);
rest.silent.interceptors.response.use(parseResponse, handleErrors);

rest.interceptors.request.use((request) => {
  startNProgress();
  return request;
});

rest.interceptors.response.use(
  (response) => {
    stopNProgress();
    return response;
  },
  (error) => {
    stopNProgress();
    return error;
  }
);

watch(
  jwt,
  (v) => {
    if (v) {
      rest.defaults.headers.common.Authorization = `Bearer ${v}`;
    } else {
      delete rest.defaults.headers.common.Authorization;
    }
  },
  { immediate: true }
);

export { rest };
