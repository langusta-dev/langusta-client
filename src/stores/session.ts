import { acceptHMRUpdate, defineStore } from 'pinia';

import { jwt, setJwt, unsetJwt } from '~/composables/jwt';

import * as api from '~/api/session';

import { useLocalProfileStore } from './localProfile';

import type { RestResponse } from '~/types/api';
import type { JwtTokenPayload } from '~/types/jwt';
import type { LogInPayload, RegisterPayload } from '~/types/session';

export const useSessionStore = defineStore('session', () => {
  const localProfileStore = useLocalProfileStore();

  const isAuth = $computed(
    () => !!jwt.value || localProfileStore.isLocalProfileEnabled
  );

  const logOut = () => {
    unsetJwt();

    if (localProfileStore.isLocalProfileEnabled) {
      localProfileStore.disableLocalProfile();
    }
  };

  const _authenticate = async (
    tokenPayloadResponse: Promise<RestResponse<JwtTokenPayload>>
  ) => {
    if (isAuth) {
      logOut();
    }

    const { data } = await tokenPayloadResponse;

    if (data?.token) {
      setJwt(data.token);
    }
  };

  const logIn = (payload: LogInPayload) => _authenticate(api.logIn(payload));

  const register = async (payload: RegisterPayload) =>
    _authenticate(api.register(payload));

  return { isAuth: $$(isAuth), logIn, logOut, register };
});

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useSessionStore, import.meta.hot));
}
