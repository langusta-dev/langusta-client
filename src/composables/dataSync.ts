import { v4 as uuid } from 'uuid';

import { useLocalProfileStore } from '~/stores/localProfile';
import { useSessionStore } from '~/stores/session';

import { isOnline } from './online';

import type { Ref } from 'vue';
import type { Editable, SynchronizableData } from '~/types/dataSync';
import type { Uuid } from '~/types/uuid';

const now = () => new Date().toString();

export const useSynchronizableArray = <T extends SynchronizableData>(
  localStorageKey: string,
  initializer: () => T[] | Promise<T[]>,
  uploader: (data: T[]) => Promise<boolean>,
  deleter: (data: Uuid[]) => Promise<boolean>,
  initialData: T[]
) => {
  const sessionStore = useSessionStore();
  const localProfileStore = useLocalProfileStore();

  const data = $(useLocalStorage<T[]>(localStorageKey, initialData));

  const dataPerId = $computed(
    () => new Map<Uuid, T>(data.map((item) => [item.id, item]))
  );

  const setData = (newData: T[]) => {
    data.splice(
      0,
      data.length,
      ...newData,

      // data created via local profile should be always preserved
      ...data.filter(({ isLocalOnly }) => isLocalOnly)
    );
  };

  let isDataReady = $ref(!sessionStore.isAuth);
  whenever(
    () => sessionStore.isAuth,
    () => {
      isDataReady = false;
    }
  );

  whenever(
    () => !isDataReady,
    async () => {
      if (sessionStore.isAuth) {
        setData(await initializer());
      }

      isDataReady = true;
    },
    { immediate: true }
  );

  const dataToUpload = $(
    useLocalStorage<T[]>(`${localStorageKey}-to-upload`, [])
  );

  const dataIdsToDelete = $(
    useLocalStorage<Uuid[]>(`${localStorageKey}-ids-to-delete`, [])
  );

  const _autosync = <U>(
    itemsRef: Ref<U[]>,
    synchronizer: (items: U[]) => boolean | Promise<boolean>
  ) => {
    watch(
      [itemsRef, isOnline],
      async () => {
        if (isOnline.value && itemsRef.value.length) {
          const isSync = await synchronizer(itemsRef.value);

          if (isSync) {
            itemsRef.value.splice(0, itemsRef.value.length);
          }
        }
      },
      { immediate: true }
    );
  };

  _autosync($$(dataToUpload), uploader);
  _autosync($$(dataIdsToDelete), deleter);

  const getById = (id: Uuid): T | null => dataPerId.get(id) || null;

  const push = (item: Editable<T>): void => {
    const newItem = {
      id: uuid(),
      createdAt: now(),
      updatedAt: now(),
      ...item,
    } as T;

    if (localProfileStore.isLocalProfileEnabled) {
      newItem.isLocalOnly = true;
    }

    data.push(newItem);

    if (!newItem.isLocalOnly) {
      dataToUpload.push(newItem);
    }
  };

  const editById = (id: Uuid, item: Editable<T>): void => {
    const oldItem = getById(id);

    if (!oldItem) {
      // eslint-disable-next-line no-console
      console.error(`Failed to modify item: ${id}`);
      return;
    }

    const newItem = {
      id,
      createdAt: oldItem.createdAt,
      updatedAt: now(),
      ...item,
    } as T;

    if (oldItem.isLocalOnly) {
      newItem.isLocalOnly = true;
    } else {
      const itemToUploadIndex = dataToUpload.findIndex(
        (item) => item.id === id
      );

      if (itemToUploadIndex !== -1) {
        dataToUpload.splice(itemToUploadIndex, 1);
      }

      dataToUpload.push(newItem);
    }

    deleteById(id);
    data.push(newItem);
  };

  const deleteById = (id: Uuid): void => {
    const index = data.findIndex((item) => item.id === id);

    if (index !== -1) {
      data.splice(index, 1);
    } else {
      // eslint-disable-next-line no-console
      console.warn(`Failed to delete item: ${id}`);
    }

    dataIdsToDelete.push(id);
  };

  return {
    state: computed(() => data),
    isReady: computed(() => isDataReady),
    isInSync: computed(() => !dataToUpload.length && !dataIdsToDelete.length),
    getById,
    push,
    editById,
    deleteById,
  };
};
