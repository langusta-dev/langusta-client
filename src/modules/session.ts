import { useSession } from '~/stores/session'

import { jwt } from '~/composables/jwt'

import type { InstallModule } from '~/types/modules'

export const install: InstallModule = ({ isClient, router }) => {
  if (!isClient) {
    return
  }

  router.beforeEach((to) => {
    if (to.meta.auth) {
      const session = useSession()

      if (!session.isAuth) {
        return '/login'
      }
    }

    return true
  })

  watch(jwt, (v) => {
    if (!v) {
      router.push('/login')
    }
  })
}
