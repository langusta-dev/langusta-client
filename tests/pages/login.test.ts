import Login from '~/pages/login.vue';

import { flushPromises, mount } from '@vue/test-utils';

import { useSessionStore } from '~/stores/session';

import * as confirm from '~/composables/confirm';

describe('login page', () => {
  it('should render', () => {
    const wrapper = mount(Login);

    const inputs = wrapper.findAll('input');
    expect(inputs).toHaveLength(2);

    const buttons = wrapper.findAll('button');
    expect(buttons).toHaveLength(2);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should log in', async () => {
    const wrapper = mount(Login);

    const inputs = wrapper.findAll('input');

    const username = 'some-username';
    const password = 'some-password';

    await inputs[0].setValue(username);
    await inputs[1].setValue(password);

    const button = wrapper.find('button');
    await button.trigger('click');

    const sessionStore = useSessionStore();

    expect(sessionStore.logIn).toHaveBeenCalledOnce();
    expect(sessionStore.logIn).toHaveBeenCalledWith({ username, password });
  });

  it('should show error and clear password input on failed login', async () => {
    const wrapper = mount(Login);

    const inputs = wrapper.findAll('input');

    await inputs[0].setValue('some-username');
    await inputs[1].setValue('some-password');

    const sessionStore = useSessionStore();

    // @ts-expect-error it's readonly
    sessionStore.isAuth = false;

    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();

    expect(sessionStore.logIn).toHaveBeenCalledOnce();

    // @ts-expect-error it's readonly
    sessionStore.isAuth = false;

    // error label
    expect(wrapper.html()).toContain('login.login_failed_error');

    // password input
    expect(inputs[1].element.value).toBe('');
    expect(inputs[1].classes()).toContain('border-error');
  });

  it('should enable local profile', async () => {
    const wrapper = mount(Login);

    const buttons = wrapper.findAll('button');

    const showConfirmSpy = vi
      .spyOn(confirm, 'showConfirm')
      .mockImplementation(() => {});

    await buttons[1].trigger('click');

    expect(showConfirmSpy).toHaveBeenCalledOnce();
    expect(showConfirmSpy).toHaveBeenCalledWith({
      msg: 'login.local_profile_confirm',
      cb: expect.any(Function),
    });
  });
});
