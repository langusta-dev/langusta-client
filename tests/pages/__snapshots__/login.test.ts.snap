// Vitest Snapshot v1

exports[`login page > should render 1`] = `
"<div _flex=\\"~ col\\" _items-center=\\"\\">
  <div _h=\\"[15%]\\"></div>
  <div _flex=\\"~ col\\" _items-center=\\"\\" _text-center=\\"\\" _gap3=\\"\\">
    <transition-group-stub enterfromclass=\\"op0\\" leavetoclass=\\"op0\\" moveclass=\\"duration-300 z-1\\" enteractiveclass=\\"pointer-events-none duration-300\\" leaveactiveclass=\\"absolute pointer-events-none duration-300\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
      <!--v-if-->
      <div>
        <div>login.username</div><input placeholder=\\"\\" type=\\"text\\" class=\\"p-x2 p-y1 w72 max-w-full bg-primary rounded placeholder:text-sm placeholder:italic placeholder:text-primary-contrast/75 border-1px outline outline-2px outline-offset-0 outline-transparent hover:outline-accent !focus:outline-accent-focus transition-colors transition-property-[outline] border-primary-contrast/30\\" autocomplete=\\"false\\" _text-center=\\"\\">
      </div>
      <div>
        <div>login.password</div><input placeholder=\\"\\" type=\\"password\\" class=\\"p-x2 p-y1 w72 max-w-full bg-primary rounded placeholder:text-sm placeholder:italic placeholder:text-primary-contrast/75 border-1px outline outline-2px outline-offset-0 outline-transparent hover:outline-accent !focus:outline-accent-focus transition-colors transition-property-[outline] border-primary-contrast/30\\" autocomplete=\\"false\\" _text-center=\\"\\">
      </div>
      <div _flex=\\"~ col\\" _items-center=\\"\\" _text-center=\\"\\" _gap2=\\"\\"><button class=\\"bg-accent-interactive p-x4 p-y.5\\" disabled=\\"\\" _rounded=\\"\\" _inline-block=\\"\\" _fw600=\\"\\" _text-accent-interactive-contrast=\\"\\" _transition-colors=\\"\\" _cursor-pointer=\\"\\" _select-none=\\"\\" _disabled:pointer-events-none=\\"\\">login.log_in</button><button class=\\"bg-accent-alt-interactive p-x3 p-y.5 text-sm\\" _rounded=\\"\\" _inline-block=\\"\\" _fw600=\\"\\" _text-accent-interactive-contrast=\\"\\" _transition-colors=\\"\\" _cursor-pointer=\\"\\" _select-none=\\"\\" _disabled:pointer-events-none=\\"\\">login.enable_local_profile</button></div>
      <div>login.has_no_account <a href=\\"/register\\" _inline-block=\\"\\" _underline=\\"\\" _fw500=\\"\\" _text-primary-contrast-interactive=\\"\\" _transition-colors=\\"\\" _cursor-pointer=\\"\\" _select-none=\\"\\">login.sign_up</a></div>
    </transition-group-stub>
  </div>
</div>"
`;
